### PR TITLE
Use Eager Code To Determine Conv Layout

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87641
* #87650
* __->__ #87305

The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -422,11 +422,53 @@ def nyi(fake_mode, func, *args, **kwargs):
...
+                padding=kwargs["padding"],
+                dilation=kwargs["dilation"],
+                transposed=kwargs["transposed"],
+                output_padding=kwargs["output_padding"],
+                groups=kwargs["groups"],
+                bias_sizes=kwargs["bias_sizes"],
+            )
+        mem_fmt = torch._C._conv_determine_backend_memory_format(
+            kwargs["input"], kwargs["weight"], conv_backend
+        )
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:452]: You don't actually care about the backend, right? Just the output memory format. Maybe only expose that?

**Review by ezyang — APPROVED** [review]

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (eellison) [reply torch/_subclasses/fake_tensor.py:452]: yea, `_select_conv_backend` was already exposed fwiw

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -422,11 +422,53 @@ def nyi(fake_mode, func, *args, **kwargs):
     assert func not in _device_not_kwarg_ops, f"NYI: {func}"
 
 
-# Meta tensors give you the ability to run PyTorch code without having to
-# actually do computation through tensors allocated on a `meta` device.
-# Because the device is `meta`, meta tensors do not model device propagation.
-# FakeTensor extends MetaTensors to also carry an additional `fake_device`
-# which tracks devices that would have been used.
+@register_op_impl(
+    lambda func: func in (aten.convolution.default, aten.convolution_backward.default)
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:426]: So the idea is that the conv meta impl in `_meta_registrations.py` is mostly correct, but gives the wrong memory format - so we manually call it and then patch up the memory format here in `FakeTensorMode`?

Is there any reason not to just patch the fancier memory format logic directly into the existing meta function? (If the answer is just that's more effort and work for later then totally fair - just want to confirm my understanding).

Comment (eellison) [reply torch/_subclasses/fake_tensor.py:426]: Most of the complexity of the logic comes in cpu, so it's not that high pri at the moment, but it was a good way to unblock turning on fake tensors for inductor shape prop.

We will eventually need to patch the fancier memory format logic yea - but there might be a way to do something similar as we're doing here even for symbolic shapes

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:426]: This logic should be moved into `_meta_registrations.py`

Comment (eellison) [reply torch/_subclasses/fake_tensor.py:426]: Yea, I did a best effort fix [here](https://github.com/pytorch/torchdynamo/pull/1639). There is a [lot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L193) [lot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L239-L273) [lot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L313-L325) of stuff to copy over.

**Pull request closed without merging**